_posts/2021-01-21-ooops-can-I-try-again-please.md

```diff
@@ -0,0 +1,73 @@
+
+> More information about errors, failed messages, and retries are available in my ["Businesses don't fail, they make mistakes"](https://milestone.topics.it/2019/09/10/businesses-dont-fail-they-make-mistakes.html) and in David Boike's excellent ["I caught an exception. Now what?"](https://particular.net/blog/but-all-my-errors-are-severe) article on Particular's blog.
+
+Is this it? Or is there something more that we can do? There are use cases in which not going through the incoming message retry process is beneficial. Moreover, there are use cases where there is no incoming message to retry, but there is only an outgoing message. Let's start with the latter.
+
+## No incoming message
+
+All systems interact with the external world, and I fail to find a scenario in which a system with no outside input makes sense. Information and requests come into the system in many different ways. We cannot expect the caller/sender of these requests to use the same retry approach previously described.
+
+> It's also one of the mantras of distributed systems design: do not offload your problems to the caller.
+
+Imagine a simple scenario in which the caller uses HTTP to send a request to the system; the system offloads the incoming request to a queue to be processed by a background service. If the offload operation, or any other step happening in between, fails, it would be nice to try to avoid merely replying with an HTTP500; it's vulgar.
+Let's imagine the problem being a transient failure in the queuing system or in the network, e.g., the sender endpoint cannot contact the message broker. If it's a transient failure, it would be nice to retry a few times before resorting to the HTTP500. In pseudo-code, it could look like:
+
+```
+public Task<HttpResponse> ProcessOrder(int orderId)
+{
+    var retryAttempts = 0;
+    while(retryAttempts < 5)
+    {
+        try
+        {
+            sendMessage(new ProcessOrder(){ OrderId = orderId });
+            return Ack();
+        }
+        catch
+        {
+            retryAttempts++;
+        }
+    }
+
+    return Fail();
+}
+```
+
+> Sending the message is not the only thing that can fail; as discussed in my last article ["Ehi! What's up? Feedback to users' requests in distributed systems"](https://milestone.topics.it/2021/01/12/feedback-to-users-requests-in-distributed-systems.html), in a complex system, there are many moving parts. In case we need to track the user request to provide feedback, the HTTT request handling code will be more complex, and more things can fail and are worth retrying.
```

Comment:
```suggestion
> Sending the message is not the only thing that can fail; as discussed in my last article ["Ehi! What's up? Feedback to users' requests in distributed systems"](https://milestone.topics.it/2021/01/12/feedback-to-users-requests-in-distributed-systems.html), in a complex system, there are many moving parts. In case we need to track the user request to provide feedback, the HTTP request handling code will be more complex, and more things can fail and are worth retrying.
```